### PR TITLE
feat: ai-cache plugin (exact-match cache for openai-chat)

### PR DESCRIPTION
**Pull request by janiussyafiq**
### Description

The **`ai-cache`** plugin (RFC #13290; supersedes #13308 / #13370): an exact-match L1 response cache for `openai-chat`, backed by Redis. Delivered as a **stack of small PRs** so each is reviewable in isolation (#13308 was closed for being too large):

① `redis-schema` shared-table fix (separate small PR) → ② **this PR** — the complete, useful exact-match cache → ③ redis-cluster support → ④ `ai-proxy-multi` semantics (draft, design question stated up front) → ⑤ RFC #13290 alignment (consumer scoping, `bypass_on`, headers schema, Prometheus metrics).

#### This PR

A correct, conservative exact-match cache with **zero edits to shared files** — all production code lives in new files (`ai-cache.lua`, `ai-cache/key.lua`, `ai-cache/schema.lua`), plus one plugin-list line each in `cli/config.lua` and `config.yaml.example`.

- **Cache key** — SHA-256 over the canonical request body **as received from the client**: every field is hashed (so any output-affecting parameter, including ones OpenAI adds later, scopes the key and can never silently cross-cache); only `stream` is excluded, because streaming requests are never cached. `temperature`/`top_p` are milli-quantised so `0.2` vs `0.2000001` don't shatter the cache; object keys are canonicalised recursively (`core.json.stably_encode`) while array order is preserved.
- **Config scoping** — the key is additionally scoped by `ctx.conf_id` + `ctx.conf_version` (the matched route/service/plugin_config identity and modification index). Any config edit — including an in-place `ai-proxy` `options.model` override change — bumps `conf_version`, so entries cached under an older config are unreachable afterwards. This covers the route/service/plugin-config + model/upstream context requested in #13308 review feedback.
- **Phase hooks** — a single `access` handler at priority **1028** (below `ai-proxy` 1040, `ai-rate-limiting` 1030, and `ai-aliyun-content-moderation` 1029, following that plugin's pattern): gate on `openai-chat`, `stream=true` → `X-AI-Cache-Status: SKIP-STREAM`, Redis `GET`, HIT short-circuits with the cached body while remaining subject to rate limiting and request moderation. The response is captured via the plugin's own `lua_body_filter`; `log` writes it via `ngx.timer.at(0)` `SETEX`.
- **`ai-proxy-multi` routes bypass caching** (merged-conf check, so service- and plugin_config-attached configs are covered): with per-instance overrides, request-as-received keying cannot distinguish instances, so PR1 restricts itself to `ai-proxy` routes where it is provably safe. Multi-instance semantics are PR ④.
- Backend: single-node `redis` (`redis-cluster` is PR ③).
- Fail-open: Redis connect/GET errors and corrupt cached entries degrade to MISS, never 5xx; corrupt entries are deleted on read.
- Cache writes only on upstream 2xx + JSON + body ≤ 1 MiB.
- Reference doc (`docs/en/latest/plugins/ai-cache.md`) + docs index entry.

**Tests** — `t/plugin/ai-cache-key.t` (key-derivation contract: determinism, canonicalisation, quantisation edge cases, signed-field regression, conf scoping/invalidation, hash-all property) and `t/plugin/ai-cache.t` (end-to-end: SKIP-STREAM, miss→hit round-trip with upstream invoked exactly once, HIT serves the stored bytes, config-edit invalidation, fail-open on Redis down, corrupt-entry drop-and-delete, `ai-proxy-multi` bypass incl. service-attached, oversize/non-2xx/non-JSON not written, TTL applied).

#### Which issue(s) this PR fixes:

Refs #13290

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible

### CODE REVIEW

File: apisix/plugins/ai-cache.lua [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    return cli, err, "single"
+end
+
+
+local function release(cli, mode, conf)
+    if mode == "cluster" then
+        -- rediscluster keeps its own pool
+        return
+    end
+    cli:set_keepalive(conf.redis_keepalive_timeout, conf.redis_keepalive_pool)
```
[the remark this answers is not in the document]
Comment (janiussyafiq) [reply apisix/plugins/ai-cache.lua]: Addressed in 8b698907 — `release()` now captures the `set_keepalive` return value and logs a warning on failure, matching the pattern in `limit-conn-redis` / `limit-count-redis`.

File: docs/en/latest/plugins/ai-cache.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+The Plugin computes a SHA-256 cache key from the request and looks it up in Redis before the request is forwarded upstream. On a **hit**, the cached response is returned directly and the upstream is never contacted. On a **miss**, the request is proxied normally and a successful JSON response is written to the cache for subsequent requests.
+
+This release implements an **exact-match cache** for the `openai-chat` protocol. All `openai-chat`-compatible providers (for example Azure OpenAI, DeepSeek, OpenRouter, Together, vLLM, Ollama) are cached transparently because they share the protocol.
+
+:::note
+
+The cache key is currently derived from the **client-sent** request body — operator-side overrides applied by `ai-proxy` (such as a `model` override) are not yet reflected in the key. If a route applies such overrides, flush the cache (or use a dedicated Redis database) before relying on cache correctness.
+
+:::
```
[the remark this answers is not in the document]
Comment (janiussyafiq) [reply docs/en/latest/plugins/ai-cache.md:54]: Addressed in 8b698907 — the plugin doc was rewritten. The note now states the key is derived from the **effective** (post-override) request, so `ai-proxy`/`ai-proxy-multi` overrides are reflected and cannot cause cross-caching.

File: docs/en/latest/plugins/ai-cache.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+| redis_timeout              | integer         | False                             | 1000    | greater than 0       | Redis connection/read/write timeout in milliseconds.                                                 |
+| redis_cluster_nodes        | array[string]   | True when `policy` is `redis-cluster` |     |                      | List of Redis cluster node addresses, used when `policy` is `redis-cluster`.                         |
+| redis_cluster_name         | string          | True when `policy` is `redis-cluster` |     |                      | Name of the Redis cluster, used when `policy` is `redis-cluster`.                                    |
+| redis_cluster_ssl          | boolean         | False                             | false   |                      | If true, use SSL to connect to the Redis cluster.                                                    |
+| redis_cluster_ssl_verify   | boolean         | False                             | false   |                      | If true, verify the Redis cluster SSL certificate.                                                   |
```
[the remark this answers is not in the document]
Comment (janiussyafiq) [reply docs/en/latest/plugins/ai-cache.md]: Addressed in 8b698907 — the attributes table now documents `redis_username`, `redis_ssl`, `redis_ssl_verify`, `redis_keepalive_timeout`, and `redis_keepalive_pool`.

File: conf/config.yaml.example [hunk 4]
```diff
@@ -519,6 +520,7 @@ plugins:                           # plugin list (sorted by priority)
   - data-mask                      # priority: 1500
   #- error-log-logger              # priority: 1091
   - proxy-cache                    # priority: 1085
+  - ai-cache                       # priority: 1086
```
[the remark this answers is not in the document]
Comment (janiussyafiq) [reply conf/config.yaml.example]: Addressed — `ai-cache` (1086) is now ordered ahead of `proxy-cache` (1085) in `conf/config.yaml.example` (8b698907), and likewise in `apisix/cli/config.lua` (5a963d98).

**Comment by janiussyafiq** [comment]
Closing — restarting the `ai-cache` plugin from a fresh, decomposed design. A new technical RFC splits the work into smaller, independently reviewable PRs (exact cache first, then semantic, streaming, observability), keyed on the request fingerprint. New PRs will follow. Thanks for the earlier review that prompted the split.
